src/tutorial/index.rst

```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
 
-We will start by implementing the agent function, whereby each agent outputs a message sharing their location.
+In FLAME GPU 2, agent functions can be implemented using the C++ :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the function's name, the function's message input style and the function's message output style. Then the function can be implemented from this, with the macro call being treated as the function's prototype.
 
-We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` is used for the output message argument.
+The C++ format of agent function description can also be used to compile agent functions specified as C++ strings at runtime. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the agent function/model has not changed). When using Python it is possible to specify agent functions using the C++ format as well as via a *pure* Python description which is shown in this tutorial. Agent function in *pure* Python must be defined as having a ``@pyflamegpu.agent_function`` decorator and using the following syntax ``def outputdata(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):`` which includes the specifiation of the name and type (uysing type annotations) of the output and input message. The *pure* Python implementation will translate the Python to C++ at runtime prior to compilation through a process known as transpiling.
+
+To describe our behaviour, we will start by implementing the agent function, whereby each agent outputs a message sharing their location.
+
+We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` (``pyflamegpu.MessageNone in Python``) is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` (``pyflamegpu.MessageSpatial2D in Python``) is used for the output message argument.
```

Comment:
in Python should be part of the inline code block x2

Also do you not mean pure-python here.

Reply:
I have called this Agent Python and made a few changes throughout the document as well as clarifying that *pure* python is a subset of Python with this name.

Reply:
See new comment, you left one in code tags.